backend/src/api/routers/projects.py

```diff
@@ -0,0 +1,138 @@
+import uuid
+from typing import Annotated, AsyncIterator
+
+import pydantic
+from api.auth.users import current_active_user
+from database.session import async_session_maker
+from fastapi import APIRouter, Depends, HTTPException, status
+from models.project import Project
+from sqlalchemy import select
+from sqlalchemy.exc import NoResultFound
+from sqlalchemy.ext.asyncio import AsyncSession
+
+router = APIRouter()
+
+
+class ReadProject(pydantic.BaseModel):
+    id: uuid.UUID
+    title: str
+    summary: str | None = None
+    status: str
+
+
+class CreateProject(pydantic.BaseModel):
+    title: str
+    summary: str | None = None
+    status: str
+
+
+class UpdateProject(pydantic.BaseModel):
+    title: str | None = None
+    summary: str | None = None
+    status: str | None = None
+
+
+async def get_project(project_id: uuid.UUID, session: AsyncSession) -> Project:
+    try:
+        project_entry = (
+            await session.execute(select(Project).filter_by(id=project_id))
+        ).scalar_one()
+    except NoResultFound:
+        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND)
+    return project_entry
+
+
+async def transactional_session() -> AsyncIterator[AsyncSession]:
+    async with async_session_maker.begin() as session:
+        yield session
+
+
+SessionWithTransactionContext = Annotated[AsyncSession, Depends(transactional_session)]
+
+
+@router.get(
+    "/",
+    response_description="List all projects",
+    dependencies=[Depends(current_active_user)],
+)
+async def list_projects(
+    session: SessionWithTransactionContext,
+) -> list[ReadProject]:
+    stmt = select(Project)
+    projects = (await session.scalars(stmt)).all()
+    return [
+        ReadProject(id=p.id, title=p.title, summary=p.summary, status=p.status)
+        for p in projects
+    ]
```

Comment:
Let's return database objects and do FastAPI the heavy lifting. Otherwise, the models are validated two times. However, this requires the `from_attributes=True` to be set in the ReadProject model (see https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.from_attributes).

```suggestion
async def list_projects(
    session: SessionWithTransactionContext,
) -> list[Project]:
    stmt = select(Project)
    return (await session.scalars(stmt)).all()
```

Reply:
Thanks that is a very good point I didn't know about. I refactored the logic a little bit by now but I'll see whether I can still take advantage of it.